option to colorize output

mohoff opened the pull request:
Donezo:

- [x] add helper `prettyColorizedSchema` to colorize the output (gray out lines starting with `#`, yellow for types, e.g. `<string>`, blue/red for optional/optional fields respectively.

No extra dependencies involved.

mohoff committed:
at option to colorize output

NorfairKing commented on yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:
```diff
@@ -38,8 +39,23 @@ prettyParserDoc = prettySchema . explainParser
+--
+-- This is meant for humans.
+-- The output may look like YAML but it is not.
+prettyColorizedSchema :: Schema -> Text
```
It's `gray` in the US but you used `grey`, so please also use `Colourise`.

mohoff replied:
renamed to gray

NorfairKing commented on yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:
```diff
@@ -57,29 +73,32 @@ emptyComments = Comments []
-              Just cd -> vsep [cd, s]
+              Just cd -> vsep [cd, annotate Yellow s]
        in \case
             EmptySchema -> e "# Nothing to parse" cs
```
This is also a comment I guess, so best to make it grey too.

mohoff replied:
done

NorfairKing commented on yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:
```diff
@@ -57,29 +73,32 @@ emptyComments = Comments []
+          e :: Doc Color -> Comments -> Doc Color
           e s cs' =
             case mkCommentsMDoc cs' of
               Nothing -> s
```
You probably want to make `s` yellow here too.

mohoff replied:
done

NorfairKing commented on yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:
```diff
@@ -38,8 +39,23 @@ prettyParserDoc = prettySchema . explainParser
+prettyColorizedSchema :: Schema -> Text
+prettyColorizedSchema = renderSimplyDecorated id startColor resetColor . layoutPretty defaultLayoutOptions . schemaDoc
+
+startColor :: Color -> Text
```
You can put `startColor` and `resetColor` in a `where` clause of `prettyColorisedSchema` because you don't need it outside of that function.

mohoff replied:
ah right, I keep forgetting that

mohoff replied:
done

NorfairKing commented on yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:
```diff
@@ -57,29 +73,32 @@ emptyComments = Comments []
+          mkCommentsMDoc :: Comments -> Maybe (Doc Color)
           mkCommentsMDoc = \case
             Comments [] -> Nothing
             Comments l -> Just $ align $ vsep $ map mkComment l
```
This is where the actual rendering of the comments happens.
I think we should keep the `Doc` inside the `Comments` constructor un-annotated, and then only annotate it here when we get them out of this `Comments` constructor.
That way definitely all comments have been annotated correctly and it's checked by the compiler.

mohoff replied:
Yeah I agree. I made some changes that should implement your suggestion

NorfairKing reviewed (commented):
Some minor comments.

I'd also be ok with colour-coding the `required` and `optional` keys (not the comments) as `red` and `blue` respectively but that's optional.

Could you also add an entry to the changelog and bump the version so that I can make a release out of this?

NorfairKing commented:
![intray](https://user-images.githubusercontent.com/3521180/82653246-35dfcc80-9c1f-11ea-9339-fa200c1895fe.png)
It seems to work when I use an incorrect config file, when I use `prettyColorizedSchema` in `YamlParse.Applicative.IO` (I'd like to make that change in this PR).
But it doesn't seem to work when I use it in in `confDescWith` in `YamlParse.Applicative.OptParse`.
I wish there was a way to get it to "just work" with the optparse stuff.

mohoff committed:
fixed visual bugs and added visual improvments

mohoff commented:
@NorfairKing Fixed visual bugs for `FieldSchema`, `AltSchema`, and `EmptySchema`. Also changed the display of `ObjectSchema` to have type `<object>`. Added blue+red for optional+required. Addressed all comments so far.

I don't know about `confDescWith` or how to fix it.

NorfairKing commented:
![colour](https://user-images.githubusercontent.com/3521180/82753729-310a4c80-9dc8-11ea-99eb-e0729d0dd0ba.png)
It's really pretty!

* About `confDescWith`, see https://github.com/pcapriotti/optparse-applicative/issues/391. It's not something for this PR because we'd have to reimplement the pretty printer for optparse.
* While we're here. Could we make the keys bright white? Now the keys have the same colour as the comments for me.
* Could you also make this change please, then we have colours automatically if something fails to parse:

```
--- a/yamlparse-applicative/src/YamlParse/Applicative/IO.hs
+++ b/yamlparse-applicative/src/YamlParse/Applicative/IO.hs
@@ -49,7 +49,7 @@ readFirstConfigFile files = go files
                         fs -> "While parsing files:" : map (("* " <>) . toFilePath) fs
                       referenceMsgs =
                         [ "Reference: ",
-                          T.unpack $ prettySchema $ explainParser (yamlSchema :: YamlParser a)
+                          T.unpack $ prettyColorizedSchema $ explainParser (yamlSchema :: YamlParser a)
```

NorfairKing commented on yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:
```diff
@@ -58,32 +77,32 @@ comment :: Text -> Comments
+            EmptySchema -> e emptyDoc $ addMComment cs $ Just "Nothing to parse"
             AnySchema -> e "<any>" cs
-            ExactSchema t -> e (pretty t) cs
+            ExactSchema t -> e (pretty t <+> annotate Gray "(exact)") cs
```
I like this change, but :

1. The colours don't work it seems:
![colour_000](https://user-images.githubusercontent.com/3521180/82753780-a1b16900-9dc8-11ea-9ed9-3f12ac97b66c.png)
2. Could we put that in a comment instead of in parentheses so that colour-blind people don't get confused:
```
e (pretty t) (cs <> comment "exact")
```
We essentially never want to show information in colour only.

mohoff replied:
Hm, for me it dims `(exact)` (blend yellow with gray) which i found to be looking nicely.

Even without colors, I find the distinction pretty clear so don't see an issue for colorblind people. But can change it for stylistic reasons if you like

NorfairKing replied:
![comment](https://user-images.githubusercontent.com/3521180/82753921-e4c00c00-9dc9-11ea-8a4a-68cab1b16e6d.png)
How do you like this?

```
            ExactSchema t ->
              let mExactCommentDoc = mkCommentsMDoc $ comment "(exact)"
               in ( case mExactCommentDoc of
                      Nothing -> id
                      Just cd -> (<+> cd)
                  )
                    $ e (pretty t) cs
```
The blending you saw is likely not portable so we shouldn't rely on it.

mohoff replied:
Done using `fromJust`

mohoff commented:
>     * While we're here. Could we make the keys bright white? Now the keys have the same colour as the comments for me.
> 

I think your color scheme is tweaked. For me, the keys are bright white and the comments grayed out. Before I try changes that I can't test, can you please either provide my a color code that reflects "bright white" in your terminal - or try yourself first?

mohoff committed:
added colored print when parsing fails

NorfairKing commented:
![white](https://user-images.githubusercontent.com/3521180/82754042-995a2d80-9dca-11ea-8831-5c2b1010ef34.png)
This is what I had in mind.

And here is my complete diff:
https://pastebin.com/prqUBVBd


The code you asked for is `"\x1b[37".

Please check what it looks like in your terminal too.

NorfairKing commented on yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:
```diff
@@ -92,17 +111,21 @@ schemaDoc = go emptyComments
             -- so they are erased when going down
-            ObjectSchema t s -> e (ge s) (addMComment cs t)
+            ObjectSchema t s -> vsep
+              [ e "<object>" (addMComment cs t)
```
Why did you add this piece? 
Now it looks like there are two pieces of an object expected:
![object](https://user-images.githubusercontent.com/3521180/82754099-05d52c80-9dcb-11ea-8ba5-2d8697d7f999.png)

mohoff replied:
In the top-level case it looks weird (although correct I think?) but if you check a nested object, I think that's an improvement. If you check `Data.Yaml.Value`, this is the correct YAML data type.

NorfairKing replied:
Ah yes, your'e right. I'll add a special case for the top level like this and revert the normal case:
```
            ObjectSchema t AnySchema -> e "<object>" (addMComment cs t)
            ObjectSchema t s -> ge s
```

NorfairKing commented:
Also please make sure the ci passes by running `nix-shell` in the root and then running `pre-commit run -a`.

mohoff committed:
make 'exact' a comment

mohoff committed:
force white color for object keys

NorfairKing commented on yamlparse-applicative/src/YamlParse/Applicative/Pretty.hs:
```diff
@@ -127,7 +128,7 @@ schemaDoc = go emptyComments
                     where blueOptional = annotate Blue "optional"
                in vsep
-                    [ keyDoc <> ":" <+> mkComment requiredDoc,
+                    [ annotate White keyDoc <> ":" <+> mkComment requiredDoc,
```
This is not the only keyDoc, see the `<key>` stuff below.

mohoff replied:
ah, I see. Done

mohoff committed:
add more white for keys

mohoff commented:
> Also please make sure the ci passes by running `nix-shell` in the root and then running `pre-commit run -a`.

I don't have a nix setup nor am I familiar with these commands. Can you please elaborate - a CONTRIBUTING.md would be sweet - generally and to help with the CI issues of this MR

NorfairKing commented:
@mohoff Don't worry then, I'll get CI working. Just ping me here when you're done.

mohoff commented:
@NorfairKing Done from my side. I left the discussions unresolved.

NorfairKing commented:
Merged manually, thanks!!
